lms/djangoapps/course_structure_api/v0/views.py

```diff
@@ -0,0 +1,227 @@
+
+        * name: The name of the course.
+
+        * uri: The URI to use to get details of the course.
+
+        * course: The course number.
+
+        * due:  The due date. For courses, the value is always null.
+
+        * org: The organization specified for the course.
+
+        * id: The unique identifier for the course.
+    """
+    paginate_by = 10
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = serializers.CourseSerializer
+
+    def get_queryset(self):
+        course_ids = self.request.QUERY_PARAMS.get('course_id', None)
+
+        course_descriptors = []
+        if course_ids:
+            course_ids = course_ids.split(',')
+            for course_id in course_ids:
+                course_key = CourseKey.from_string(course_id)
+                course_descriptor = courses.get_course(course_key)
+                course_descriptors.append(course_descriptor)
+        else:
+            course_descriptors = modulestore().get_courses()
+
+        results = [course for course in course_descriptors if self.user_can_access_course(self.request.user, course)]
+
+        # Sort the results in a predictable manner.
+        results.sort(key=lambda x: x.id)
+
+        return results
+
+
+class CourseDetail(CourseViewMixin, RetrieveAPIView):
+    """
+    **Use Case**
+
+        CourseDetail returns details for a course.
+
+    **Example requests**:
+
+        GET /{course_id}/
+
+    **Response Values**
+
+        * category: The type of content.
+
+        * name: The name of the course.
+
+        * uri: The URI to use to get details of the course.
+
+        * course: The course number.
+
+        * due:  The due date. For courses, the value is always null.
+
+        * org: The organization specified for the course.
+
+        * id: The unique identifier for the course.
+    """
+
+    serializer_class = serializers.CourseSerializer
+
+    def get_object(self, queryset=None):
+        return self.get_course_or_404()
+
+
+class CourseStructure(CourseViewMixin, RetrieveAPIView):
+    """
+    **Use Case**
+
+        Retrieves course structure.
+
+    **Example requests**:
+
+        GET /{course_id}/structure/
+
+    **Response Values**
+
+        * root: ID of the root node of the structure
+
+        * blocks: Dictionary mapping IDs to block nodes.
+    """
+    serializer_class = serializers.CourseStructureSerializer
+
+    def get_object(self, queryset=None):
+        # Make sure the course exists and the user has permissions to view it.
+        course = self.get_course_or_404()
+
+        try:
+            # Check for existing course structure
+            course_structure = models.CourseStructure.objects.get(course_id=course.id)
+        except models.CourseStructure.DoesNotExist:
+            # If we don't have data stored, generate it.
+            course_structure = models.update_course_structure(course.id)
```

Comment:
We have no explicit SLAs for the endpoints. The alternatives are to either return a 404 and be done, or trigger the generation task asynchronously and return a 503 with a Retry-After header.